tools/codegen/schema/typed_model.go

```diff
@@ -0,0 +1,114 @@
+package schema
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/mongodb/terraform-provider-mongodbatlas/tools/codegen/codespec"
+)
+
+func GenerateTypedModels(attributes codespec.Attributes) CodeStatement {
+	return generateTypedModels(attributes, "TFModel", false)
+}
+
+func generateTypedModels(attributes codespec.Attributes, name string, isNested bool) CodeStatement {
+	models := []CodeStatement{generateStructOfTypedModel(attributes, name)}
+
+	if isNested {
+		models = append(models, generateModelObjType(attributes, name))
+	}
+
+	for i := range attributes {
+		additionalModel := getAdditionalModelIfNested(&attributes[i])
+		if additionalModel != nil {
+			models = append(models, *additionalModel)
+		}
+	}
+
+	result := GroupCodeStatements(models, func(list []string) string { return strings.Join(list, "\n") })
+	return result
+}
+
+func generateModelObjType(attrs codespec.Attributes, name string) CodeStatement {
+	structProperties := []string{}
+	for i := range attrs {
+		propType := attrModelType(&attrs[i])
+		prop := fmt.Sprintf(`%q: %sType,`, attrs[i].Name.SnakeCase(), propType)
+		structProperties = append(structProperties, prop)
+	}
+	structPropsCode := strings.Join(structProperties, "\n")
+	return CodeStatement{
+		Code: fmt.Sprintf(`var %sObjType = types.ObjectType{AttrTypes: map[string]attr.Type{
+    %s
+}}`, name, structPropsCode),
+		Imports: []string{"github.com/hashicorp/terraform-plugin-framework/types", "github.com/hashicorp/terraform-plugin-framework/attr"},
+	}
+}
+
+func getAdditionalModelIfNested(attribute *codespec.Attribute) *CodeStatement {
```

Comment:
should this just be `getNestedModel()` ?

Reply:
agree, adjusted